src/toxcore/messenger/packet/file_send_request.rs

```diff
@@ -0,0 +1,140 @@
+/*! FileSendRequest struct.
+It is used to start transferring file to a friend.
+*/
+
+use nom::{rest, le_u8, be_u32, be_u64};
+
+use std::str;
+
+use crate::toxcore::binary_io::*;
+use crate::toxcore::crypto_core::*;
+
+/// Type of file to transfer
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum FileType {
+    /// Normal data file.
+    Data = 0,
+    /// Avatar image file.
+    Avatar,
+}
+
+/// Maximum file name size in bytes
+const MAX_FILESEND_FILENAME_LENGTH: usize = 255;
+
+impl FromBytes for FileType {
+    named!(from_bytes<FileType>,
+        switch!(be_u32,
+            0 => value!(FileType::Data) |
+            1 => value!(FileType::Avatar)
+        )
+    );
+}
+
+/** FileSendRequest is a struct that holds info to start transferring file to a friend.
+
+This packet is used to start transferring sender's file to a friend.
+`file_type` and `file_size` are sent in big endian format.
+
+Serialized form:
+
+Length    | Content
+--------- | ------
+`1`       | `0x50`
+`1`       | `file_id`
+`4`       | `file_type`(0 = normal file, 1 = avatar file)
+`8`       | `file_size`
+`32`      | `file_unique_id`(a PrecomputedKey)
```

Comment:
Where did you get that?

```c
       case PACKET_ID_FILE_SENDREQUEST: {
            const unsigned int head_length = 1 + sizeof(uint32_t) + sizeof(uint64_t) + FILE_ID_LENGTH;

            if (data_length < head_length) {
                break;
            }

            uint8_t filenumber = data[0];

#if UINT8_MAX >= MAX_CONCURRENT_FILE_PIPES

            if (filenumber >= MAX_CONCURRENT_FILE_PIPES) {
                break;
            }

#endif

            uint64_t filesize;
            uint32_t file_type;
            uint16_t filename_length = data_length - head_length;

            if (filename_length > MAX_FILENAME_LENGTH) {
                break;
            }

            memcpy(&file_type, data + 1, sizeof(file_type));
            file_type = net_ntohl(file_type);

            memcpy(&filesize, data + 1 + sizeof(uint32_t), sizeof(filesize));
            net_to_host((uint8_t *) &filesize, sizeof(filesize));
            struct File_Transfers *ft = &m->friendlist[i].file_receiving[filenumber];

            if (ft->status != FILESTATUS_NONE) {
                break;
            }

            ft->status = FILESTATUS_NOT_ACCEPTED;
            ft->size = filesize;
            ft->transferred = 0;
            ft->paused = FILE_PAUSE_NOT;
            memcpy(ft->id, data + 1 + sizeof(uint32_t) + sizeof(uint64_t), FILE_ID_LENGTH);

            VLA(uint8_t, filename_terminated, filename_length + 1);
            uint8_t *filename = nullptr;

            if (filename_length) {
                /* Force NULL terminate file name. */
                memcpy(filename_terminated, data + head_length, filename_length);
                filename_terminated[filename_length] = 0;
                filename = filename_terminated;
            }

            uint32_t real_filenumber = filenumber;
            real_filenumber += 1;
            real_filenumber <<= 16;

            if (m->file_sendrequest) {
                (*m->file_sendrequest)(m, i, real_filenumber, file_type, filesize, filename, filename_length,
                                       userdata);
            }

            break;
        }
```

Also file name is missing

Reply:
`file_id` in c-toxcore is `file_unique_id` in tox-rs and `file_number` in c-toxcore is `file_id` in tox-rs.

>Also file name is missing
What do you mean?

Reply:
In my first version I used `file_number` but later changed to `file_id` by comments.
I don't mind to change it to `file_number` or other.

Reply:
https://github.com/TokTok/c-toxcore/blob/0aad180d1e1d7dae7cbf2868e139af0a4dc34fe7/toxcore/tox.c#L51-L53

Reply:
If we return to `file_number` then we must chage other `file related` ser/de which are using `file_id` as `file_number`.

Reply:
There is:

```
            uint16_t filename_length = data_length - head_length;

            if (filename_length > MAX_FILENAME_LENGTH) {
                break;
            }
```

What about that?

Reply:
In c-toxcore if file_name length is longer than 255 bytes then silently ignore the packet.
But it would be better that consider it as Error and print error_msg and ignore the packet.

Reply:
I see.